modules/cli/apply/apply.go

```diff
@@ -90,18 +92,39 @@ func Apply(ctx context.Context, client *client.Getter, name string, baseDirs []s
 						return nil
 					}
 				}
-				if err := Dir(ctx, client, name, path, opts); err == ErrNoResources {
+				if err := Dir(ctx, client, name, path, opts, gitRepoBundlesMap); err == ErrNoResources {
 					logrus.Warnf("%s: %v", path, err)
 					return nil
 				} else if err != nil {
 					return err
 				}
 				foundBundle = true
+
 				return nil
 			})
 			if err != nil {
 				return err
 			}
+			//list all bundles for this gitrepo and prune those not found
+			c, err := client.Get()
+			if err != nil {
+				return err
+			}
+			filter := labels.Set(map[string]string{fleet.RepoLabel: name})
+			bundles, err := c.Fleet.Bundle().List(client.Namespace, metav1.ListOptions{LabelSelector: filter.AsSelector().String()})
+			if err != nil {
+				return err
+			}
+
+			for _, bundle := range bundles.Items {
+				if ok := gitRepoBundlesMap[bundle.Name]; !ok {
+					logrus.Debugf("Bundle to be deleted since it is not found in gitrepo %v anymore %v %v", name, bundle.Namespace, bundle.Name)
+					err := c.Fleet.Bundle().Delete(bundle.Namespace, bundle.Name, nil)
+					if err != nil {
+						return err
+					}
+				}
+			}
```

Comment:
One small change: if an output is provided, I think we should only log something to stderr and perform no action. If an output is not provided, then this looks fine.

i.e. we should follow the logic in `Dir` for whether or not to run the `save` command:
https://github.com/rancher/fleet/blob/17c46aca876eabd32f1564251b24e3bdbf4e43ef/modules/cli/apply/apply.go#L192-L196

Can we also wrap the code here in a `prune` helper function, similar to `save`?

Reply:
yes, added a helper function that figures out the bundles to prune and then either deletes them or logs them to output depending on opts.Output provided